template/script.sh.erb

```diff
@@ -23,9 +23,7 @@ module list
 
 # Benchmark info
 echo "TIMING - Starting jupyter at: $(date)"
-export container_folder=${CONTAINER_REPO}/${COURSE}
-## when you test new installs you probaby want to drop the images locally in your home folder
-#export container_folder=$HOME/Programs/containers
+export container_folder=${CONTAINER_REPO}/${COURSE}/${LAST_DIGIT}
```

Comment:
I'm not sure I understand the purpose of `${LAST_DIGIT}` here. I see that it's getting the last digit of the user's canvas ID and using that as part of the container folder name. So for example:

```
$ echo "u_1234_g_98765" | sed 's/\(.*\)\([0-9]\)\(_g_.*\)/\2/' 
4
```

Do we need that? What if multiple users have the same last digit?

Reply:
That is for limiting potential number of concurrent reads from the same file.

This class could have , according to what @joshuagetega  said, up to 300 students. 
So - and this is probably is an overkill - , instead of having just 1 copy of the container, I created 10 
```
[root@holy7c26201 cs109b]# pwd
/n/singularity_images/Academic-cluster/Spring_2021/cs109b
[root@holy7c26201 cs109b]# ls ? 
0:
harvardat_cs109_fbb3898-2021-01-25-97fc43a2370f.sif  harvardat_cs109_fbb3898.sif

1:
harvardat_cs109_fbb3898-2021-01-25-97fc43a2370f.sif  harvardat_cs109_fbb3898.sif

2:
harvardat_cs109_fbb3898-2021-01-25-97fc43a2370f.sif  harvardat_cs109_fbb3898.sif

3:
harvardat_cs109_fbb3898-2021-01-25-97fc43a2370f.sif  harvardat_cs109_fbb3898.sif

4:
harvardat_cs109_fbb3898-2021-01-25-97fc43a2370f.sif  harvardat_cs109_fbb3898.sif

5:
harvardat_cs109_fbb3898-2021-01-25-97fc43a2370f.sif  harvardat_cs109_fbb3898.sif

6:
harvardat_cs109_fbb3898-2021-01-25-97fc43a2370f.sif  harvardat_cs109_fbb3898.sif

7:
harvardat_cs109_fbb3898-2021-01-25-97fc43a2370f.sif  harvardat_cs109_fbb3898.sif

8:
harvardat_cs109_fbb3898-2021-01-25-97fc43a2370f.sif  harvardat_cs109_fbb3898.sif

9:
harvardat_cs109_fbb3898-2021-01-25-97fc43a2370f.sif  harvardat_cs109_fbb3898.sif
```

and each student will end up loading from the file that corresponds to their last digit. 
So even if all of them had to start a job at the same time we would have no more than 30 concurrent reads from the same file. 
Again, likely an overkill, but it does not cost much to do. Did that last week for data fest and worked well.

The disadvantage is that we are not taking advantage of local cache on the reads, but that's ok I guess. We cannot have everything.

The best solution would be to drop a copy of the containers locally on each of the compute nodes, ... but that is harder to maintain.

Reply:
Thanks for explaining @pontiggi - I hadn't thought about it like that (e.g. having 10 copies of the image to distribute reads). I'm not sure how well the user IDs will be distributed for that scheme, but I suppose it doesn't have to be a perfect distribution to see the benefits. If it worked well for datafest, then I imagine this will work pretty well for this course and is easier to maintain than copying to the compute nodes. Seems like a reasonable trade-off.